Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -525,56 +542,59 @@ void FrameLoader::submitForm(Ref<FormSubmission>&& submission)
         // Continuing form submission makes no sense in the latter case.
         // There is a repeat check after timer fires, so this is not a correctness issue.
 
-        targetFrame = &m_frame;
+        targetFrame = frame.copyRef();
     } else
         submission->clearTarget();
 
     if (!targetFrame->page())
         return;
 
-    if (m_frame.tree().isDescendantOf(targetFrame))
+    if (frame->tree().isDescendantOf(targetFrame.get()))
         m_submittedFormURL = submission->requestURL();
 
     submission->setReferrer(outgoingReferrer());
-    submission->setOrigin(SecurityPolicy::generateOriginHeader(m_frame.document()->referrerPolicy(), submission->requestURL(), m_frame.document()->securityOrigin(), OriginAccessPatternsForWebProcess::singleton()));
+    submission->setOrigin(SecurityPolicy::generateOriginHeader(frame->document()->referrerPolicy(), submission->requestURL(), m_frame->document()->securityOrigin(), OriginAccessPatternsForWebProcess::singleton()));
 
     targetFrame->navigationScheduler().scheduleFormSubmission(WTFMove(submission));
 }
 
 void FrameLoader::stopLoading(UnloadEventPolicy unloadEventPolicy)
 {
-    if (m_frame.document() && m_frame.document()->parser())
-        m_frame.document()->parser()->stopParsing();
+    // The frame may have started destruction, in which case we can't protect it.
+    RefPtr protectedFrame = m_frame->view() ? m_frame.ptr() : nullptr;
+
+    if (m_frame->document() && m_frame->document()->parser())
+        m_frame->document()->protectedParser()->stopParsing();
 
     if (unloadEventPolicy != UnloadEventPolicy::None)
         dispatchUnloadEvents(unloadEventPolicy);
 
     m_isComplete = true; // to avoid calling completed() in finishedParsing()
     m_didCallImplicitClose = true; // don't want that one either
 
-    if (m_frame.document() && m_frame.document()->parsing()) {
+    if (m_frame->document() && m_frame->document()->parsing()) {
```

Comment:
Can we store document in a local variable here?
if (RefPtr document = m_frame->document(); document && document->parsing()) {
    finishedParsing();
    document->setParsing(false);
}